activesupport/lib/active_support/buffered_logger.rb

```diff
@@ -63,6 +63,11 @@ def add(severity, message = nil, progname = nil, &block)
       # If a newline is necessary then create a new message ending with a newline.
       # Ensures that the original message is not mutated.
       message = "#{message}\n" unless message[-1] == ?\n
+      # We should encode message to encoding appropriate for writing to log file,
+      #   which is `Encoding.default_internal` because log file is created without any encoding options.
+      # Otherwise we receive exception during buffer flush stage
+      #   which will occur forever in current rails process, for all ongoing requests.
+      message = message.encode( Encoding.default_internal, :invalid => :replace, :undef => :replace )      
```

Comment:
This is unfortunatelty not a safe way to do it:
- `Encoding.default_internal` is not always set. I guess `Encoding.default_external` might work, but I'm not a 100% sure that would always be correct either.
- The code needs to stay compatible with Ruby 1.8.7.

Rails 3.x seems to have solved the issue by treating log messages as binary (see https://github.com/rails/rails/commit/8f999a3f80284e08804c374cb5690c8bc158e4c1). To me this seems like the prefered solution since it stays close to the behaviour in Ruby 1.8.7. Will have to adapt the solution a bit though, since BufferedLogger was changed in Rails 3.x.

Reply:
Thank you for the clarifications!